File: 力扣/学习计划/算法入门/024_字母大小写全排列/main.java
import java.util.ArrayList;
import java.util.List;

/**
 * https://leetcode-cn.com/problems/letter-case-permutation/
 * 字母大小写全排列
 */
class Solution {
    List<String> res = new ArrayList<>();

    public List<String> letterCasePermutation(String s) {
        dfs(s.toCharArray(), 0);
        return res;
    }

    /**
     * @param cs    搜索的字符数组
     * @param start 开始搜索的位置
     */
    void dfs(char[] cs, int start) {
        res.add(String.valueOf(cs));
        for (int i = start; i < cs.length; i++) {
            // 数字, 则跳过
            if (Character.isDigit(cs[i])) continue;
            // 大小写反转
            cs[i] = reverseLetterCase(cs[i]);
            // 搜索
            dfs(cs, i + 1);
            // 回溯, 大小写反转回来
            cs[i] = reverseLetterCase(cs[i]);
        }
    }

    /**
     * 反转大小写
     * 'A' --> 'a'
     * 'a' --> 'A'
     */
    public char reverseLetterCase(char c) {
        return Character.isUpperCase(c) ? Character.toLowerCase(c) : Character.toUpperCase(c);
    }
}

File: 力扣/学习计划/算法入门/024_字母大小写全排列/main.ts
function letterCasePermutation(s: string): string[] {
    let res = new Array()

    const dfs = function (s: string, begin: number) {
        res.push(s)
        for (let i = begin; i < s.length; i++) {
            // 数字, 跳过此轮循环
            if (isDigit(s[i])) continue;
            // 字母, 反转字母大小写
            s = s.substr(0, i) + changeLetter(s[i]) + s.substr(i + 1)
            // 搜索
            dfs(s, i + 1)
            // 回溯, 将字母大小写反转回来
            s = s.substr(0, i) + changeLetter(s[i]) + s.substr(i + 1)
        }
    }

    dfs(s, 0)
    return res
};

/**
 * 反转字母大小写
 */
const changeLetter = function (s: String): String {
    return s >= 'a' && s <= 'z' ? s.toUpperCase() : s.toLowerCase()
}

/**
 * 判断是否是数字
 */
const isDigit = function (s: String): Boolean {
    return s >= '0' && s <= '9'
}